File: LeetCode/200.Number_of_Islands/200.Number_of_Islands.java
class Solution {

    int[][] dirs = { { 1, 0 }, { 0, 1 }, { -1, 0 }, { 0, -1 } };

    public int numIslands(char[][] grid) {
        int m = grid.length, n = grid[0].length;
        boolean[][] visited = new boolean[m][n];

        int res = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j] && grid[i][j] == '1') {
                    dfs(grid, visited, i, j);
                    res++;
                }
            }
        }
        return res;
    }

    private void dfs(char[][] grid, boolean[][] visited, int i, int j) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || visited[i][j] || grid[i][j] == '0') {
            return;
        }

        visited[i][j] = true;
        for (int[] d : dirs) {
            dfs(grid, visited, i + d[0], j + d[1]);
        }
    }
}

File: LeetCode/200.Number_of_Islands/200.Number_of_Islands.swift
class Solution {
    func numIslands(_ grid: [[Character]]) -> Int {
        var grid = grid
        let m = grid.count
        guard m > 0 else { return 0 }
        let n = grid[0].count
        let dirs = [(1, 0), (0, 1), (-1, 0), (0, -1)]

        var q: Deque<(Int, Int)> = []

        var res = 0
        for i in 0 ..< m {
            for j in 0 ..< n {
                if grid[i][j] == "1" {
                    res += 1
                    grid[i][j] == "0"
                    q.append((i, j))
                    while let (x, y) = q.popFirst() {
                        for (dx, dy) in dirs {
                            let nx = dx + x, ny = dy + y
                            if nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == "1" {
                                q.append((nx, ny))
                                grid[nx][ny] = "0"
                            }
                        }
                    }
                }
            }
        }

        return res
    }
}
